File: system-test/src/testkit/dsl/core/usecase/usecases/CancelOrder.ts
import type { BookShopDriver } from '../../../../driver/port/book-shop-driver.js';
import type { CancelOrderResponse } from '../../../../driver/port/dtos/CancelOrderResponse.js';
import { UseCaseResult } from '../../shared/use-case-result.js';
import { VoidVerification } from '../../shared/void-verification.js';
import type { UseCaseContext } from '../../shared/use-case-context.js';
import { BaseBookShopUseCase } from './base/BaseBookShopUseCase.js';

export class CancelOrder extends BaseBookShopUseCase<CancelOrderResponse, VoidVerification> {
  private _orderNumberResultAlias: string = '';

  constructor(driver: BookShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  orderNumber(orderNumberResultAlias: string): this {
    this._orderNumberResultAlias = orderNumberResultAlias;
    return this;
  }

  async execute(): Promise<UseCaseResult<CancelOrderResponse, VoidVerification>> {
    const orderNumber = this.context.getResultValue(this._orderNumberResultAlias) ?? this._orderNumberResultAlias;
    const result = await this.driver.cancelOrder({ orderNumber });

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}

File: system-test/src/main/java/com/jcupac/bookshop/testkit/dsl/core/usecase/usecases/CancelOrder.java
package com.jcupac.bookshop.testkit.dsl.core.usecase.usecases;

import com.jcupac.bookshop.testkit.driver.port.BookShopDriver;
import com.jcupac.bookshop.testkit.driver.port.dtos.CancelOrderRequest;
import com.jcupac.bookshop.testkit.dsl.core.usecase.usecases.base.BaseBookShopUseCase;
import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseResult;
import com.jcupac.bookshop.testkit.dsl.core.shared.UseCaseContext;
import com.jcupac.bookshop.testkit.dsl.core.shared.VoidVerification;

public class CancelOrder extends BaseBookShopUseCase<Void, VoidVerification> {
    private String orderNumberResultAlias;

    public CancelOrder(BookShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public CancelOrder orderNumber(String orderNumberResultAlias) {
        this.orderNumberResultAlias = orderNumberResultAlias;
        return this;
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var orderNumber = context.getResultValue(orderNumberResultAlias);
        var request = CancelOrderRequest.builder().orderNumber(orderNumber).build();
        var result = driver.cancelOrder(request).mapVoid();
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}
